File: codex-agent-runtime-desktop/bindings/python/consumer/smoke.py
import inspect
import typing

from codex_agent import (
    ApprovalPreset,
    AuthenticationState,
    AuthenticationStatus,
    AuthorizationUrl,
    Capability,
    ClientInfo,
    CollaborationMode,
    CodexAgent,
    CodexAuthentication,
    CodexConversation,
    CodexConnectors,
    CodexHooks,
    CodexHost,
    CodexIntegrationAuthorization,
    CodexInteractions,
    CodexMcpServers,
    CodexModels,
    CodexPlugins,
    CodexSkills,
    ConversationId,
    Conversations,
    ConversationValue,
    Elicitation,
    ElicitationAction,
    ElicitationResponse,
    FormField,
    FormFieldType,
    FormTextListValue,
    FormTextValue,
    HOST_STATE_NEW,
    HostState,
    HostStateKind,
    HostStateReady,
    Message,
    MessageRole,
    PlanProgress,
    PlanStep,
    PlanStepStatus,
    PluginReference,
    TurnRequest,
    McpHttpTransport,
    McpServer,
    McpServerConfiguration,
    McpAuthStatus,
    McpToolApproval,
    InteractionState,
    PendingApproval,
    StateStream,
    TurnProgress,
    Workspace,
)


assert tuple(
    service.__name__
    for service in (
        CodexAuthentication,
        CodexConnectors,
        CodexHooks,
        CodexIntegrationAuthorization,
        CodexInteractions,
        CodexMcpServers,
        CodexModels,
        CodexPlugins,
        CodexSkills,
    )
) == (
    "CodexAuthentication",
    "CodexConnectors",
    "CodexHooks",
    "CodexIntegrationAuthorization",
    "CodexInteractions",
    "CodexMcpServers",
    "CodexModels",
    "CodexPlugins",
    "CodexSkills",
)


client = ClientInfo("consumer", "Python consumer", "1.0")
assert client.name == "consumer"
assert ConversationId("abc").value == "abc"
assert ApprovalPreset.ASK_ME.value == 2
assert AuthenticationStatus.AUTHENTICATED.value == 2
assert HostStateKind.READY.value == 4
assert McpToolApproval.WRITES.value == 2

headers = {"X-Client": "consumer"}
transport = McpHttpTransport("https://example.com/mcp", headers=headers)
configuration = McpServerConfiguration("consumer", transport, enabled_tools=[])
server = McpServer("consumer", "Consumer", McpAuthStatus.OAUTH, configuration)
headers["X-Client"] = "changed"
assert server.is_authorized
assert server.configuration.transport.headers["X-Client"] == "consumer"
assert server.configuration.enabled_tools == ()
assert server.configuration.disabled_tools is None

steps = [PlanStep("compile", PlanStepStatus.COMPLETED)]
progress = PlanProgress("done", steps)
steps.clear()
assert progress.steps == (PlanStep("compile", PlanStepStatus.COMPLETED),)
assert PluginReference("id", "plugin", "market").uri == "plugin://plugin@market"

values = ["one", "one", "two"]
field = FormField(
    "items",
    "Items",
    FormFieldType.MULTI_SELECT,
    default_value=FormTextListValue(values),
)
values.clear()
assert field.default_value == FormTextListValue(("one", "one", "two"))
assert AuthenticationState(AuthenticationStatus.SIGNED_OUT).failure is None
assert Message(
    "message",
    MessageRole.USER,
    "hello",
    capabilities={Capability.WEB_SEARCH},
).capabilities == frozenset({Capability.WEB_SEARCH})
assert (
    TurnRequest("hello", collaboration_mode=CollaborationMode.PLAN).collaboration_mode
    is CollaborationMode.PLAN
)
assert HOST_STATE_NEW is HOST_STATE_NEW
assert issubclass(HostStateReady, HostState)
assert inspect.signature(HostStateReady).parameters["agent"].annotation == "CodexAgent"
assert inspect.iscoroutinefunction(CodexHost.start)
assert inspect.iscoroutinefunction(CodexHost.select_workspace)
assert inspect.iscoroutinefunction(CodexHost.aclose)
host_state = inspect.getattr_static(CodexHost, "state")
assert isinstance(host_state, property)
assert typing.get_type_hints(host_state.fget)["return"] == StateStream[HostState]

assert inspect.iscoroutinefunction(Conversations.read)
assert typing.get_type_hints(Conversations.read)["return"] is ConversationValue
assert inspect.iscoroutinefunction(CodexConversation.send)
assert typing.get_type_hints(CodexConversation.send)["prompt"] == str | TurnRequest
assert isinstance(CodexConversation.current_messages, property)
assert (
    typing.get_type_hints(CodexConversation.current_messages.fget)["return"]
    == StateStream[tuple[Message, ...]]
)
assert isinstance(CodexConversation.active_turn_progress, property)
assert (
    typing.get_type_hints(CodexConversation.active_turn_progress.fget)["return"]
    == StateStream[TurnProgress | None]
)

for name, return_type in {
    "authentication": CodexAuthentication,
    "connectors": CodexConnectors,
    "conversations": Conversations,
    "hooks": CodexHooks,
    "integration_authorization": CodexIntegrationAuthorization,
    "interactions": CodexInteractions,
    "mcp_servers": CodexMcpServers,
    "models": CodexModels,
    "plugins": CodexPlugins,
    "skills": CodexSkills,
    "workspace": Workspace,
}.items():
    member = inspect.getattr_static(CodexAgent, name)
    assert isinstance(member, property)
    assert typing.get_type_hints(member.fget)["return"] is return_type

form_field = FormField(
    "name",
    "Name",
    FormFieldType.STRING,
    is_required=True,
    default_value=FormTextValue("Codex"),
)
elicitation = Elicitation(
    "request",
    ConversationId("conversation"),
    "server",
    "Choose",
    form=(form_field,),
)
content = elicitation.initial_values()
assert form_field.accepts(content["name"])
assert elicitation.validate(content).is_valid
response = elicitation.accept(content)
assert elicitation.accepts(response)
assert ElicitationResponse.decline().action is ElicitationAction.DECLINE
assert ElicitationResponse.cancel().action is ElicitationAction.CANCEL

approval = PendingApproval(
    "request", ConversationId("conversation"), "Title", "Details"
)
interactions = InteractionState((approval, approval), {"request"})
assert interactions.is_resolving(approval)
assert interactions.pending_for(ConversationId("conversation")) == (approval, approval)

chat_gpt_url = AuthorizationUrl.chat_gpt("https://auth.openai.com/authorize")
external_url = AuthorizationUrl.external("http://localhost:8787/callback")
assert "auth.openai.com" not in repr(chat_gpt_url)
assert external_url.value.startswith("http://localhost:")

File: codex-agent-runtime-desktop/bindings/dart/consumer/bin/smoke.dart
import 'package:codex_agent/codex_agent.dart';

Future<void> main() async {
  final client = CodexClientInfo(
    name: 'consumer',
    title: 'Dart consumer',
    version: '1.0.0',
  );
  final cancellation = CodexCancellation();
  final request = CodexTurnRequest(
    prompt: 'hello',
    capabilities: const <CodexCapability>{CodexCapability.webSearch},
    invocations: const <CodexInvocation>[
      CodexSkillInvocation(name: 'review', path: 'skills/review.md'),
    ],
  );
  final mcp = CodexMcpServerConfiguration(
    name: 'local_tools',
    transport: CodexMcpStdioTransport(command: 'tool'),
  );
  if (client.name != 'consumer' ||
      cancellation.isCancelled ||
      request.invocations.single.key != 'skill:skills/review.md' ||
      mcp.name != 'local_tools') {
    throw StateError('public Dart package contract is unusable');
  }

  final field = CodexFormField(
    name: 'name',
    title: 'Name',
    type: CodexFormFieldType.string,
    isRequired: true,
    defaultValue: const CodexTextFormValue('Codex'),
  );
  final elicitation = CodexElicitation(
    requestId: 'request',
    serverName: 'server',
    conversationId: CodexConversationId('conversation'),
    message: 'Name',
    form: [field],
  );
  final content = <String, CodexFormValue>{
    'name': const CodexTextFormValue('Codex'),
  };
  final accepted = elicitation.accept(content);
  final validation = elicitation.validate(content);
  if (!field.accepts(content['name']) ||
      elicitation.initialValues().keys.single != 'name' ||
      validation.issues.isNotEmpty ||
      !elicitation.accepts(accepted) ||
      CodexElicitationResponse.cancel().action !=
          CodexElicitationAction.cancel ||
      CodexElicitationResponse.decline().action !=
          CodexElicitationAction.decline) {
    throw StateError('synchronous elicitation projection is unusable');
  }

  final approval = CodexPendingApproval(
    requestId: 'approval',
    conversationId: CodexConversationId('conversation'),
    title: 'Approve',
    details: 'Details',
  );
  final interactions = CodexInteractionState(
    pending: [approval, approval],
    resolvingRequestIds: ['approval'],
  );
  final selected = interactions.pendingFor(CodexConversationId('conversation'));
  if (!interactions.isResolving(approval) ||
      selected.length != 2 ||
      !identical(selected.first, approval) ||
      CodexAuthorizationUrl.chatGpt('https://auth.openai.com/authorize')
              .purpose !=
          CodexAuthorizationPurpose.chatGpt ||
      CodexAuthorizationUrl.external('http://localhost/callback').purpose !=
          CodexAuthorizationPurpose.external) {
    throw StateError('synchronous interaction projection is unusable');
  }
}

// Compiled from the installed package so every leaf projection remains a
// public, statically typed Dart API.
Future<void> compileLeafServiceSurface({
  required CodexAuthentication authentication,
  required CodexInteractions interactions,
  required CodexIntegrationAuthorization authorization,
  required CodexModels models,
  required CodexSkills skills,
  required CodexHooks hooks,
  required CodexPlugins plugins,
  required CodexConnectors connectors,
  required CodexMcpServers mcpServers,
  required CodexModel model,
  required CodexPluginReference plugin,
  required CodexHook hook,
  required CodexSkill skill,
  required CodexMcpServer server,
  required CodexMcpServerConfiguration configuration,
  required CodexPendingApproval approval,
  required CodexPendingElicitation elicitation,
  required CodexIntegration integration,
}) async {
  await authentication.authenticate(CodexApiKeyAuthentication('key'));
  await authentication.cancel();
  await authentication.signOut();
  await authentication.state.current;
  authentication.state.changes.listen((_) {});
  await authentication.isAuthenticatedState.current;
  authentication.isAuthenticatedState.changes.listen((_) {});
  await authentication.isAuthenticatingState.current;
  authentication.isAuthenticatingState.changes.listen((_) {});

  await interactions.openUrl(elicitation);
  await interactions.resolveApproval(approval, CodexApprovalDecision.accept);
  await interactions.resolveElicitation(
    elicitation,
    CodexElicitationResponse(action: CodexElicitationAction.cancel),
  );
  await interactions.state.current;
  interactions.state.changes.listen((_) {});
  await interactions.approvalsState.current;
  interactions.approvalsState.changes.listen((_) {});
  await interactions.elicitationsState.current;
  interactions.elicitationsState.changes.listen((_) {});

  await authorization.authorize(integration);
  await authorization.cancel();
  await authorization.state.current;
  authorization.state.changes.listen((_) {});
  await authorization.activeState.current;
  authorization.activeState.changes.listen((_) {});
  await authorization.isAuthorizingState.current;
  authorization.isAuthorizingState.changes.listen((_) {});

  await models.list();
  await models.resolve();
  await models.resolveEffort(model);
  await models.resolveServiceTier(model);
  await skills.list();
  await skills.read('/skill');
  await skills.install('/skill', CodexInstallationScope.user);
  await skills.uninstall(skill);
  await hooks.list();
  await hooks.install('/hook', CodexInstallationScope.user);
  await hooks.trust(hook);
  await hooks.uninstall(hook);
  await plugins.list();
  await plugins.read(plugin);
  await plugins.install(plugin);
  await plugins.uninstall(plugin);
  await connectors.list();
  await mcpServers.list();
  await mcpServers.add(configuration);
  await mcpServers.remove(server);
  final availability = <bool>[
    skills.isAvailable,
    hooks.isAvailable,
    plugins.isAvailable,
    connectors.isAvailable,
    mcpServers.isAvailable,
  ];
  if (availability.every((value) => value)) return;
}

void compileAgentSurface(CodexAgent agent) {
  final projections = <Object>[
    agent.authentication,
    agent.connectors,
    agent.conversations,
    agent.hooks,
    agent.integrationAuthorization,
    agent.interactions,
    agent.mcpServers,
    agent.models,
    agent.plugins,
    agent.skills,
  ];
  final workspace = agent.workspace;
  if (projections.length != 10 || workspace.path.isEmpty) {
    throw StateError('public Agent projection is unusable');
  }
}

void compileHostSurface(CodexHost host, CodexAgent agent) {
  final create = CodexHost.create;
  final ready = CodexReadyHostState(agent);
  final lifecycle = host.state;
  final start = host.start;
  final selectWorkspace = host.selectWorkspace;
  final close = host.close;
  if (!identical(ready.agent, agent) ||
      ready.kind != CodexHostStateKind.ready ||
      <Object>[create, lifecycle, start, selectWorkspace, close].length != 5) {
    throw StateError('public Host/READY projection is unusable');
  }
}
